### PR TITLE
Add Dependent Haskell quantifiers (including the new "foreach")

### PR DESCRIPTION
**Pull request by goldfirere**
This proposal reserves syntax for all the quantifiers of Dependent Haskell, along with introducing the `-XDependentTypes` extension and allowing the new quantifiers to be used in very limited contexts. This proposal does *not* actually add dependent types.

[Rendered](https://github.com/goldfirere/ghc-proposals/blob/pi/proposals/0000-pi.rst)

### CODE REVIEW

**Comment by nomeata** [comment]
It seems that some aspects of `-Wcompat` are enabled only when `-XDependentTypes`, but some compat warnings are introduced even without the extension enabled. Intentional?

**Comment by goldfirere** [comment]
Yes, that was intentional. But even reading it now, I'm not sure it's the best design. The question is about whether, say, `data Proxy :: forall k. k -> Type` should be accepted in the distant future. That `forall` really should be a `pi`. What makes `pi` a keyword, though? If it's `-XDependentTypes`, do we want to require someone using kind polymorphism (but not other dependent types) to turn on the extension? Do we just add `pi` to `-XPolyKinds`? I don't know. The proposal takes the conservative approach of allowing the `Proxy` declaration if `-XDependentTypes` isn't turned on.

**Comment by simonpj** [comment]
> Here, a "kind" is one that can be syntactically recognized as such, by appearing to the right of a :: in a type

This looks highly dubious to me.  Proposal #103 is all about trying to avoid such artificial syntactic distinctions.  Eg what about a visible kind argument -- is that "syntactically recognised as a kind"?

**Comment by goldfirere** [comment]
According to the proposal, a visible kind argument is not "syntactically recognized as a kind", and I'm not changing that stance. However, you're right that this distinction smells bad. I don't know a better solution, though.

Right now, we use `->` to mean two different things: sometimes, it's matchable and sometimes, it's unmatchable. In Dependent Haskell, `->` uniquely means unmatchable, and `'->` (subject to bikeshedding, if anyone cares to) uniquely means matchable. (Type inference includes a sub-typing relationship between these for convenience. See chapter 6 of my thesis if you dare.)

The problem is that we want to introduce the `'->` syntax to allow for backward compatibility later. We also ideally would like to encourage authors to use the new syntax where it should be used: in kinds. This proposal includes a clumsy heuristic (but, I wager, an effective one) for doing so. We could omit this part of the proposal (specifically, point 4 under **Semantics:**) without great harm.

Do you see a better way to do this?

**Comment by Saagar-A** [comment]
Bikeshedding: I don't like where the `'` is located in the matchable versions of the dependent quantifiers. To me, it makes more sense for them to be associated with the `->` or `.`. Looking something like this:
```Haskell
forall a '-> ...
forall a '. ...
pi a '->
pi a '.
```
This way there is a symmetry between `a -> b` : `forall x -> b` and `a '-> b` : `forall x '-> b`

**Comment by goldfirere** [comment]
@Saagar-A Yes, I too noticed the lack of parallel between `'->` and `'pi x ->`. The one problem with your suggested syntax is that the tick will be confused with a prime. For example, what would `forall a'.` mean? Is that binding the variable `a'` (a-prime)? Or is it binding `a` in a matchable `forall`? Presumably we would use whitespace to tell the difference, which would work fine, in practice. At this point, I still prefer my syntax, but I'll list your suggestion as an alternative.

**Comment by Ericson2314** [comment]
@goldfirere I like that alternative; I'd straight-up allow `'` in infix operators across the board to enshrine this without another whitespace hack.

**Comment by andrewthad** [comment]
I also prefer the alternative syntax for the matchable dependent quantifiers.

**Comment by andrewthad** [comment]
The proposal mentions:

1. With `-XDependentTypes`, the matchable arrow becomes *required* in kind signatures.
2. With `-XDependentTypes`, matchable arrows and relevant quantification becomes *allowed* (but optional) in the kind signatures of a data type.

One thing that I'm unclear on is whether or not the kind signature on a data type declaration requires a matchable arrow. The proposal says:

> As described above in the part introducing matchability data constructor arguments are matchable (as are type constructor arguments). Thus, they should be able to use the matchable arrow '->.

This seems weaker than the later statement:

> The matchable arrow '-> is required (with -XDependentTypes) in kinds

More concretely, my question is whether or not this will be allowed when `-XDependentTypes` is on:

```
data Foo :: Bool -> Type where
  FooC :: Foo 'True
```

Is the matchable arrow required here or not? Personally, I would like to see stricter, rather than looser, requirements here. In fact, I'd really prefer that `pi` be *required*, instead of merely *suggested*, in data type kind signatures when `-XDependentTypes` is on. After all, `-XDependentHaskell` already carries with it the matchable kind signature requirement, so full compatibility with code written without the extension is already out the window. As someone who frequently reads other people's code, I find it easier to understand when people provide as many type signatures and kind signatures as possible. More stringent requirements would may make things a little more difficult for authors, but they would benefit readers by making data type kind signatures more reflective of the actual kind.

**Comment by goldfirere** [comment]
I've updated the proposal with @andrewthad's suggestion about matchability, which I don't fully agree with. Type and data constructor types are so common, and matchability is so esoteric, that I don't want to burden everyone with requiring the matchability annotations. That said, I would support an optional warning that would encourage the pedantic among us to put in the matchability marks.

**Comment by andrewthad** [comment]
I like the idea of a `-Wpedantic-matchability` flag. +1 from me on that.

**Comment by glaebhoerl** [comment]
(:bike: warning)

The `pi` keyword seems awfully jargony to me. Is there no better keyword (or syntax) we could potentially use here?

Consider the connections one has to be able to make to derive from "pi" its intended meaning:

 1. "pi" stands for the greek letter π
 2. Uppercase Π is used in math to denote products
 3. Because they are conceptually products over the inhabitants of the input type, type theory uses Π as one of its notations for dependent functions
 4. Ah, so it's a dependent function

Compare this to `forall`:

 1. `forall a. a -> a` means  "for all types `a`, a function from `a` to `a`". It reads right off.

(I know that I, personally, was pretty mystified when I first heard about "pi types" many years ago. What the heck does the circumference of a circle have to do with types?)

Given that Haskell probably has more mainstream penetration than all dependently typed languages combined, it seems likely that `-XDependentTypes` is going to be most peoples' first introduction to dependently typed programming, and we can't really expect them to be already familiar with all of the relevant background.

**Comment by cartazio** [comment]
There’s also how will this avoid breaking use of pi the geometry constant.
Or is that already addressed?

**Comment by int-index** [comment]
Interesting concern raised by @glaebhoerl. My personal favorite to go with `forall` would be `foreach`. It makes sense that `forall` is parametric (we have the same behavior _for all_ instantiations), whereas `foreach` is not (we may pattern match and define an individual behavior _for each_ instantiation).

**Comment by glaebhoerl** [comment]
I had that thought as well, but thought the distinction between "forall" and "foreach" might be too subtle and therefore also surprising. But if other people don't hate it, it seems like a decent alternative to me.

As another possibility, is there any grammatical obstacle to simply leaving off the keyword (or, to preserve symmetry, we could claim that it's Ɛ) and requiring parentheses, for the usual dependent function syntax?

E.g.

```haskell
-- parametric
forall (n :: Nat) -> Vec n -> Vec n

-- dependent
(n :: Nat) -> Vec n -> Vec n 
```

Hmm... I just had this GHCi session:

```haskell
$ ghci
GHCi, version 8.0.2: http://www.haskell.org/ghc/  :? for help
Prelude> :set -XDataKinds -XPolyKinds
Prelude> data Nat = Z | S Nat
Prelude> data Vec (n :: Nat) = Vec
Prelude> let foo :: (n :: Nat) -> Vec n -> Vec n; foo = undefined

<interactive>:5:12: error:
    • Expected a type, but ‘n’ has kind ‘Nat’
    • In the type signature:
        foo :: (n :: Nat) -> Vec n -> Vec n

Prelude> let bar :: (a :: *) -> Maybe a -> Maybe a; bar a = id
Prelude> 
```
So that might be a no-go. It's... strange that this syntax is already accepted?

**Comment by goldfirere** [comment]
The syntax that @glaebhoerl suggests directly above is taken, unfortunately. For example, I can say `id :: (a :: Type) -> a`, where `:: Type` is a kind signature. Note that this is a different meaning than `pi`!

@glaebhoerl thanks for the bikeshedding: I had never considered anything but `pi` for the new keyword, but you make a compelling argument that `pi` is quite hard to motivate if you're not a type theorist. I have to say that I really like `foreach`. It's nicely parallel to `forall`, and the meaning is just about right. (And, @cartazio, I hadn't thought about `pi = 3.14`. That would indeed be a problem.) It will be hard to retrain my fingers at this point, but I'll update the proposal (and title of the ticket).

**Comment by andrewthad** [comment]
My preference would be `pi` instead of `foreach`. I'm not worried about the name conflict with the mathematical constant since people could just use another name for it in code where they need it. Ultimately, the committee will have to pick decide what name is used, but I just wanted to make sure that there's one voice out there in favor of `pi`.

The mediocre argument I can offer is an argument from precedent. Haskell tends to pick mathematical names for abstractions rather than intuitive names. Every now and then, I see a discussion on the libraries list about why `Functor` wasn't named `Mappable` (and similar concerns around `Applicative`, `Monad`, `Monoid`, `Semigroup`, etc.). The advantages are that it makes it easier to discuss the thing or to learn about what it means outside the context of haskell.

**Comment by chrisdone** [comment]
@andrewthad I think you can make a balancing scale like this:

* You use the theory name:
   1. It'll be intuitive to the people who know the theory. 
   2. There's an established body of literature about that name. On the other hand, that literature usually doesn't directly apply to this particular implementation of it.
   2. You'll have to describe what it "means" in every day programming terms in the documentation. And sometimes in dozens of blog posts.
* You use the "every day" name:
   1. It'll be intuitive to newbies.
   2. There won't be a literature/google results about that name. On the other hand, if your implementation is well documented, this won't be a problem.
   3. People who know the theory will have to look at the docs to discover that "it's really just a ..."

If the scales are equal, I'd just go with the "every day" name to reduce suffering. 

Incidentally, I think "applicative" is an exception to the rule there -- it's for "pure function applied to funny arguments" and as far as I'm aware was made up by Conor McBride with that in mind, from the function `ap` named for the same reason. I think "Foldable" and "Traversable" are examples of similar naming convention. "Functor" etc. are the ones with obscure etymology far removed from every day words.

**Comment by vagarenko** [comment]
Perhaps that's because I'm not a native English speaker, but to me `forall` and `foreach` seem to mean the same thing.

@andrewthad 
>My preference would be pi instead of foreach. I'm not worried about the name conflict with the mathematical constant since people could just use another name for it in code where they need it. 

Why there is supposed to be a name conflict? Term `pi` is already defined in `Floating` typeclass in the Prelude. How can you put it into the type signature where it would cause conflict?

@chrisdone 
>It'll be intuitive to newbies.

IMO, seeing `forall` in one signature and `foreach` in another is not intuitive for newbies.

Are there other options besides `foreach` and `pi`?
Perhaps `dep` short for dependent, or `rel` short for relevant?

**Comment by andrewthad** [comment]
@vagarenko I am a native English speaker, and `forall` and `foreach` mean the same thing to me as well. Also, I think you're probably right about the naming conflict with the mathematical constant not actually being an issue. Also, I feel like `dep` would be a bad name since `forall` and `pi` are actually **both** dependent quantifiers. Something like `rel` or `relevant` would be better.

The more I think about it, the more I feel that `foreach` will not make things easier to understand for anyone. In most languages, that keyword implies iteration, which is not what the relevant quantifier does. I don't really think any choice of keyword will make things easier since the end user is basically has to understand what relevance is. Given this, I'd rather pick a keyword that doesn't have as much historical baggage attached to it like `rel`, `relevant`, or `pi`.

(There is some precedent here. The name of `return` in `Monad` is often considered a mistake. It was named that way to help people coming from other languages, but instead it just means that explanations of using a monadic interface always have to include a disclaimer that `return` has a misleading name. Years down the road, the `pure` method from `Applicative` was named more appropriately. No one ever has to provide a disclaimer about "pure not doing what you think it does", since people don't have any expectation about what `pure` should do).

**Comment by goldfirere** [comment]
I'm also a native English speaker, and `forall` and `foreach` have a subtle distinction for me: the first implies that the things being iterated over are somehow more tightly connected, which is the right implication here.

The point about `foreach` not being iteration is well taken. I still like it, though.

And I do see a problem with the mathematical constant. Under `-XDependentTypes`, `pi` would be a keyword. Otherwise, suppose we had

```
pi :: Type -> Type
pi = ...

x :: pi a -> a
```

Are we `pi`-quantifying `a`? Or are we calling the function `pi` on an implicitly quantified `a`? We'll never know.

**Comment by andrewthad** [comment]
Ah, I had forgotten about function promotion. The conflict is real. Also, I had forgotten that `pi` was in `Prelude`, which makes the conflict much worse than I originally imagined.

**Comment by Tritlo** [comment]
I am not a native English speaker, but for me, `fo rall` implies that it doesn't matter what the element in question is, it holds for all of them, but `for each` implies that it holds for all of them, but it matters (i.e. is relevant) what the element in question. I also quite like the rhyming in `forall` and `foreach`, whereas `pi` or `relevant` would make me think that the two were quite unlike each other.

**Comment by chrisdone** [comment]
I'm not that big on `foreach`, either. Good arguments against it. (Being a subtle distinction against `forall`, and conflicting with other language ideas.) But seeing as `pi`, is tricky as a keyword, it seems that both choices aren't obviously clear winners. 

Is `pitype` not viable? I suppose in the end, Haskellers when talking about it, will call it "pi types" even if it has syntax like `foreach`. That makes me tilt on the scales closer to the original naming. One thing `pitype` has is google-ability.

**Comment by int-index** [comment]
> Being a subtle distinction against `forall`

The actual distinction between the features is equally subtle.

**Comment by goldfirere** [comment]
Conversation here seems to have ebbed. @nomeata, I would like to submit this to the committee. Thanks!

**Comment by RyanGlScott** [comment]
Apologies for asking questions about proposals that have already been submitted to the committee, but a couple of questions about this feature popped into my mind recently.

1. Currently, there's a `forall`-or-nothing rule (see https://github.com/ghc-proposals/ghc-proposals/pull/103) which states that any type signature headed by an explicit `forall` must explicitly quantify all of its type variables. How will the `forall`-or-nothing rule evolve with the addition of these new quantifiers? Will type signatures like these:

   ```haskell
   f :: forall a -> b -> Either a b
   g :: foreach a. a -> b -> Either a b
   h :: foreach a -> b -> Either a b
   ```

   be rejected for not explicitly quantifying `b`?

2. How does this feature interact with `TypeApplications`? Currently, one can use `TypeApplications` on type variables that were bound with the `forall.` quantifier. Will this property extend to all of the other quantifiers introduced in this proposal? A subset of them?

3. GHC currently has a warning when you quantify a type variable with `forall.` but don't use it anywhere, such as in:

   ```haskell
   foo :: forall a. Int
   foo = 42
   ```
   ```
   warning: [-Wunused-foralls]
       Unused quantified type variable ‘a’
       In the type ‘forall a. Int’
   ```

   Do you think that a similar warning is warranted when you use a dependent quantifier but don't use it anywhere? My understanding is that this:

   ```haskell
   bar :: foreach a -> Int
   bar x = x
   ```

   could just have well been given the type signature `Int -> Int`, since `a` is not actually used in a dependent fashion anywhere in the type signature (save for its binding site). Note that saying `a` is "unused" would be inaccurate here, since removing `a` would change the type, but perhaps GHC could warn that `foreach a` could be changed to `Int`.

  A similar question applies to visible `forall`s.

4. Will GHC throw an error message if I try to `forall`-quantify something relevantly? In other words, can GHC infer the relevance of each variable and check that the user-provided relevance matches up with the inferred relevance?

   Conversely, if I use `foreach` irrelevantly, will GHC warn me?

**Comment by goldfirere** [comment]
Great questions! Thanks for asking.

1. I think all those should be rejected as not obeying the `forall`-or-nothing rule.

2. `TypeApplications` can be used to bind a `forall`- or `foreach`-quantified invisible variable. That is, it will work with both `forall a.` and `foreach a.` (and their matchable forms). It will thus be seen more as an invisibility-override than a type-application marker.

3. Yes, we should warn for dependently-quantified variables never used in a dependent way. But I'm unsold as to where this warning should go in GHC's warning hierarchy. Naming parameters is useful documentation, and other languages sometimes require users to do this, even when the names are redundant. (I'm looking at you, Java interfaces.) Even where these aren't required, they are allowed (C/C++ header files, all those dependently typed languages).

4. A `forall`-quantified variable used relevantly is a type error, pure and simple. A `foreach`-quantified variable not used relevantly should induce a warning, yes. This latter case can't happen, though, under this proposal, as `foreach` is very restricted. I don't think either of these cases can happen with this proposal, though.

**Comment by goldfirere** [comment]
I went to update the proposal with my answers above, but there seems to be no need. I don't think any of these cases can actually arise with the restrictions in this proposal and more properly belong in some future proposal that introduces dependent types properly.

**Comment by rrnewton** [comment]
In the bike shedding department, I didn’t see any reactions to @chrisdone’s suggestion of “pitype”. It’s googlable and one letter shorter than foreach. When you see it for the first time it kind of begs you to google it.

**Comment by glaebhoerl** [comment]
(The word "pi" would bother me less if we were actually following precedent by adopting it, but as far as I'm aware, none of the major dependently typed languages use it in their actual syntax.)

**Comment by Centril** [comment]
@glaebhoerl I think that is a factor of that in Agda and Idris, dependent types were there from the start, and so they have the nice syntax for dependent types where you just say `{A: Set} -> (x: A) -> (n: Nat) -> Vec A n`. But unfortunately we can't use `(n: Nat) -> ..` as discussed above. I'm not familiar with the syntax in Coq. I'm curious, are there any languages which have added dependent types after the fact, and if so, what did they do / use as syntax?

**Comment by int-index** [comment]
> After some reflection, my view on `pi`, `foreach` and `forall` is that the two latter ones are too close together to not cause confusion between dependent and parametric modes.

1. I disagree with the assumption that this confusion will be caused by naming — in general, people don't have the right intuition for relevance/irrelevance and its implications, so the confusion will be there either way. By using similar names we are more honest with the programmer, not trying to present those language features as completely different things.

2. Would you agree that using similar keywords for similar concepts is more appropriate than using completely different keywords for similar concepts? Pondering the difference between the words `forall` and `foreach` can provide deeper intuition about the actual distinction between those quantifiers, while thinking about the difference between `forall` and `pitype` ends quickly as "they're not similar at all".

In other words, when I imagine explaining dependent types to a colleague, I feel more comfortable saying "`foreach` is like `forall` but relevant" than "`pitype` is like `forall` but relevant", as in the former case there's a neat parallel between the concepts and their names, while in the latter case only the concepts are subtly different, but not the names.

We have a similar case with `data` type declarations, type families, and data families — the name `data family` is neatly a hybrid between the names `type family` and `data`, so I can explain data families as "injective & generative type families" while still maintaining that each data instance is just like a regular `data` declaration.

**Comment by goldfirere** [comment]
The committee has decided to table this proposal until there is a more crying need. The main motivation for writing this was to put the syntax in #81 in context, just to make sure we weren't painting ourselves into a corner. Even without acceptance, the existence of this proposal shows that #81 is forward-compatible. Thus this proposal will be marked dormant until we're closer to actually having dependent types.

**Comment by davemenendez** [comment]
I agree the arguments against `pi` as the syntax for relevant quantifiers. In general, keywords should suggest the name of a thing rather than the name of the syntax used in some other context. (I dislike Python's `lambda` keyword for the same reason.)

I like `foreach`, because it suggests a similarity to `forall` while having the sense of treating things individually instead of collectively. However, if the similarity is judged too confusing, I suggest `given` as an alternative, as it suggests having access to something. For example, `given k. k -> Type`.

**Comment by ollef** [comment]
This is perhaps tangential, but is there anywhere I can read more about matchability in Haskell, and especially the story for it in dependent Haskell? Several dependently typed systems have had inconsistencies in the past due to assuming injectivity of type constructors (see Idris issue https://github.com/idris-lang/Idris-dev/issues/3687, which links to similar issues in Agda and Lean). Is there some special sauce that makes it work out in Haskell?

**Comment by adamgundry** [comment]
@ollef Haskell's term and type languages are already inconsistent as logics (because of general recursion and type-in-type, for starters). Injectivity of type constructors merely gives another way to write looping terms/types, but we have several ways already!

We'd be in trouble (type safety would be broken) if the coercion language was inconsistent, but the coercion language is already much more restricted than the type or term languages.

**Comment by Ericson2314** [comment]
Can explicit matchablity in type families replace data families? (I didn't see this discussed in the thesis, which surprised me.)

**Comment by goldfirere** [comment]
The syntax in my thesis doesn't allow for this: type families are implemented via lambdas, which are always unmatchable. I'm not sure precisely how data families would fit in, but I'm sure there's a way.

**Comment by Ericson2314** [comment]
This is stretch from how things work today, and I believe what your thesis proposes, but say given:
```haskell
class A x where
  type Foo x :: SomeK
class B x where
  data Bar x :: SomeK
```
we have top level signatures:
```haskell
Foo :: forall x. A x => x -> SomeK
Bar :: forall x. A x '=> x '-> SomeK
```

**Comment by goldfirere** [comment]
You're right that data families are matchable and that type families aren't, but there are still several problems:

- You're assuming constrained type families here. It's a fine assumption, but it's different than what we have today (and nothing in my thesis suggests it).
- The types you write aren't quite right. I would write

    ```
    Foo :: forall k. forall (x :: k) -> A x => SomeK
    Bar :: forall k '. forall (x :: k) '-> A x '=> SomeK
    ```

    though I'm still not quite sure all the quantifiers are matchable in the data family case. Would need to think a bit harder.
- There is no way to implement `Bar` based directly on features in my thesis. I don't think.

**Comment by goldfirere** [comment]
The [Constrained Type Families paper](https://repository.brynmawr.edu/cgi/viewcontent.cgi?article=1075&context=compsci_pubs) has the details, but not in the context of dependent types.

**Comment by nomeata** [comment]
Closing this, as a dormant proposal with little activity. Re-opening is always possible.

**Comment by goldfirere** [comment]
Please see also https://github.com/ghc-proposals/ghc-proposals/pull/102#issuecomment-381467606, where I describe that the committee has broadly agreed with this proposal, but that we shouldn't implement it until other parts fall into place. Expect this to be reopened in due course!

**Comment by AntC2** [comment]
Re #236, [this comment and disucssion up-thread](https://github.com/ghc-proposals/ghc-proposals/pull/236#issuecomment-497991777) mentions possible syntax for using Visible Dependent Quantification in terms. That's at the usage site rather than the definition/quantifying site, so not directly in scope of this PR.
